state/pruner.go

```diff
@@ -0,0 +1,277 @@
+// configured pruner interval, and re-evaluates the retain height.
+type Pruner struct {
+	service.BaseService
+	logger log.Logger
+	// DB to which we save the retain heights
+	bs BlockStore
+	// State store to prune state from
+	stateStore Store
+	interval   time.Duration
+}
+
+type prunerConfig struct {
+	interval time.Duration
+}
+
+func defaultPrunerConfig() *prunerConfig {
+	return &prunerConfig{
+		interval: config.DefaultPruningInterval,
+	}
+}
+
+type PrunerOption func(*prunerConfig)
+
+// PrunerInterval allows control over the interval between each run of the
+// pruner.
+func PrunerInterval(t time.Duration) PrunerOption {
+	return func(p *prunerConfig) { p.interval = t }
+}
+
+func NewPruner(stateStore Store, bs BlockStore, logger log.Logger, options ...PrunerOption) *Pruner {
+	cfg := defaultPrunerConfig()
+	for _, opt := range options {
+		opt(cfg)
+	}
+	p := &Pruner{
+		bs:         bs,
+		stateStore: stateStore,
+		logger:     logger,
+		interval:   cfg.interval,
+	}
+	p.BaseService = *service.NewBaseService(logger, "Pruner", p)
+	return p
+}
+
+func (p *Pruner) OnStart() error {
+	go p.pruningRoutine()
+	return nil
+}
+
+// SetApplicationRetainHeight sets the application retain height with some
+// basic checks on the requested height.
+//
+// If a higher retain height is already set, we cannot accept the requested
+// height because the blocks might have been pruned.
+//
+// If the data companion has already set a retain height to a higher value we
+// also cannot accept the requested height as the blocks might have been
+// pruned.
+func (p *Pruner) SetApplicationRetainHeight(height int64) error {
+	if height <= 0 || height < p.bs.Base() || height > p.bs.Height() {
+		return ErrInvalidHeightValue
+	}
+	currentAppRetainHeight, err := p.stateStore.GetApplicationRetainHeight()
+	if err != nil {
+		if err == ErrKeyNotFound {
+			currentAppRetainHeight = height
+		} else {
+			return err
+		}
+	}
+	currentCompanionRetainHeight, err := p.stateStore.GetCompanionBlockRetainHeight()
+	var noCompanionRetainHeight bool
+	if err != nil {
+		if err == ErrKeyNotFound {
+			noCompanionRetainHeight = true
+		} else {
+			return err
+		}
+	}
+	if currentAppRetainHeight > height || (!noCompanionRetainHeight && currentCompanionRetainHeight > height) {
+		return errors.New("cannot set a height lower than previously requested - blocks might have already been pruned")
+	}
+	err = p.stateStore.SaveApplicationRetainHeight(height)
+	return err
+}
+
+// SetCompanionRetainHeight sets the application retain height with some basic
+// checks on the requested height.
+//
+// If a higher retain height is already set, we cannot accept the requested
+// height because the blocks might have been pruned.
+//
+// If the application has already set a retain height to a higher value we also
+// cannot accept the requested height as the blocks might have been pruned.
+func (p *Pruner) SetCompanionRetainHeight(height int64) error {
```

Comment:
This API uses signed integers for height, whereas in ADR-101 protobufs, the height fields are specified as `uint64` and this type is emitted in the generated gRPC interface in #1154. To avoid tedious checking logic everywhere, can we agree on a single integer type? Cc @thanethomson

This might be too late to ask, as I see both `int64` and `uint64` used for height throughout the various proto packages. Sometimes both in the same package!